File: sdk/src/main/java/io/antelli/sdk/model/Question.java
package io.antelli.sdk.model;

import android.os.Bundle;
import android.os.Parcel;
import android.os.Parcelable;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;

/**
 * Handcrafted by Štěpán Šonský on 28.08.2017.
 */

public class Question implements Parcelable {

    private static final String PARAM_QUERY = "QUERY";
    private static final String PARAM_LANGUAGE = "LANGUAGE";

    private Bundle params = new Bundle();

    public Question(@NonNull String query, @NonNull String language) {
        setQuery(query.toLowerCase());
        setLanguage(language);
    }

    public boolean equals(@NonNull String string) {
        return getQuery().equals(string.toLowerCase());
    }

    public boolean equalsOne(@NonNull String[] strings) {
        for (String string : strings) {
            if (getQuery().equals(string)) {
                return true;
            }
        }
        return false;
    }

    public boolean contains(@NonNull String string) {
        return getQuery().contains(string.toLowerCase());
    }

    public boolean containsOne(@NonNull String... strings) {
        for (String string : strings) {
            if (getQuery().contains(string.toLowerCase())) {
                return true;
            }
        }
        return false;
    }

    public boolean containsAll(@NonNull String... strings) {
        for (String string : strings) {
            if (!getQuery().contains(string.toLowerCase())) {
                return false;
            }
        }
        return true;
    }

    public boolean containsWord(@NonNull String word) {
        String input = addSpacePadding(getQuery());
        return input.contains(addSpacePadding(word.toString()));
    }

    public boolean containsOneWord(@NonNull String... words) {
        String input = addSpacePadding(getQuery());
        for (String word : words) {
            if (input.contains(addSpacePadding(word.toLowerCase()))) {
                return true;
            }
        }
        return false;
    }

    public boolean containsAllWords(@NonNull String... words) {
        String input = addSpacePadding(getQuery());
        for (String word : words) {
            if (!input.contains(addSpacePadding(word))) {
                return false;
            }
        }
        return true;
    }

    public boolean startsWith(@NonNull String prefix) {
        return getQuery().startsWith(prefix.toLowerCase());
    }

    public String removeWords(@NonNull String... words) {
        String result = addSpacePadding(getQuery());

        for (String word : words) {
            result = result.replace(addSpacePadding(word), " ");
        }
        return result.trim();

    }

    public String removeWords(@NonNull String[]... words) {
        String result = addSpacePadding(getQuery());

        for (String[] wordArray : words) {
            for (String word : wordArray) {
                result = result.replace(addSpacePadding(word), " ");
            }
        }
        return result.trim();
    }

    public String[] getWords() {
        return getQuery().split(" ");
    }

    public String getWordAt(int position) {
        String[] words = getWords();
        if (position < words.length) {
            return words[position];
        }
        return null;
    }

    public int getWordPosition(String word) {
        String[] words = getWords();
        for (int i = 0; i < words.length; i++) {
            if (words[i].equals(word)) {
                return i;
            }
        }
        return -1;
    }

    public int[] getNumbers() {
        List<Integer> list = new ArrayList<>();
        for (String word : getWords()) {
            try {
                int number = Integer.parseInt(word);
                list.add(number);
            } catch (NumberFormatException e) {
                //continue
            }
        }
        int[] result = new int[list.size()];
        for (int i = 0; i < list.size(); i++) {
            result[i] = list.get(i);
        }
        return result;
    }

    private void setQuery(String lowerCase) {
        params.putString(PARAM_QUERY, lowerCase);
    }

    public String getQuery() {
        return params.getString(PARAM_QUERY);
    }

    public void setLanguage(String language) {
        params.putString(PARAM_LANGUAGE, language);
    }

    public String getLanguage() {
        return params.getString(PARAM_LANGUAGE);
    }

    public void setParam(String name, String value) {
        params.putString(name, value);
    }

    public String getParam(String name) {
        return params.getString(name);
    }

    private String addSpacePadding(String string) {
        return " " + string + " ";
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeBundle(this.params);
    }

    protected Question(Parcel in) {
        this.params = in.readBundle(getClass().getClassLoader());
    }

    public static final Creator<Question> CREATOR = new Creator<Question>() {
        @Override
        public Question createFromParcel(Parcel source) {
            return new Question(source);
        }

        @Override
        public Question[] newArray(int size) {
            return new Question[size];
        }
    };
}

File: sdk/src/main/java/io/antelli/sdk/model/Question.kt
package io.antelli.sdk.model

import android.os.Bundle
import android.os.Parcel
import android.os.Parcelable
import java.util.*

/**
 * Handcrafted by Štěpán Šonský on 28.08.2017.
 */
class Question : Parcelable {
    private var params = Bundle()

    var query: String
        get() = params.getString(PARAM_QUERY)!!
        private set(query) {
            params.putString(PARAM_QUERY, query)
        }

    var language: String
        get() = params.getString(PARAM_LANGUAGE)!!
        set(language) {
            params.putString(PARAM_LANGUAGE, language)
        }

    val words: List<String>
        get() {
            return query.split(" ")
        }

    val numbers : List<Int>
        get() {
            val list: MutableList<Int> = ArrayList()
            for (word in words) {
                try {
                    val number = word.toInt()
                    list.add(number)
                } catch (e: NumberFormatException) {
                    //continue
                }
            }
            return list
        }

    constructor(query: String, language: String) {
        this.query = query
        this.language = language
    }

    fun equals(string: String): Boolean {
        return query.equals(string, true)
    }

    fun equalsOne(strings: Array<String>): Boolean {
        for (string in strings) {
            if (query == string) {
                return true
            }
        }
        return false
    }

    operator fun contains(string: String): Boolean {
        return query.contains(string, true)
    }

    fun containsOne(vararg strings: String): Boolean {
        for (string in strings) {
            if (query.contains(string, true)) {
                return true
            }
        }
        return false
    }

    fun containsAll(vararg strings: String): Boolean {
        for (string in strings) {
            if (!query.contains(string, true)) {
                return false
            }
        }
        return true
    }

    fun containsWord(word: String): Boolean {
        val input = addSpacePadding(query)
        return input.contains(addSpacePadding(word))
    }

    fun containsOneWord(vararg words: String): Boolean {
        val input = addSpacePadding(query)
        for (word in words) {
            if (input.contains(addSpacePadding(word), true)) {
                return true
            }
        }
        return false
    }

    fun containsAllWords(vararg words: String): Boolean {
        val input = addSpacePadding(query)
        for (word in words) {
            if (!input.contains(addSpacePadding(word), true)) {
                return false
            }
        }
        return true
    }

    fun startsWith(prefix: String): Boolean {
        return query.startsWith(prefix, true)
    }

    fun getWordAt(position: Int): String? {
        val words = words
        return if (position < words.size) {
            words[position]
        } else null
    }

    fun getWordPosition(word: String): Int {
        val words = words
        for (i in words.indices) {
            if (word == words[i]) {
                return i
            }
        }
        return -1
    }

    fun removeWords(vararg words: String): String {
        var result = addSpacePadding(query)
        for (word in words) {
            result = result.replace(addSpacePadding(word), " ")
        }
        return result.trim { it <= ' ' }
    }

    fun removeWords(vararg words: Array<String>): String {
        var result = addSpacePadding(query)
        for (wordArray in words) {
            for (word in wordArray) {
                result = result.replace(addSpacePadding(word), " ")
            }
        }
        return result.trim(' ')
    }

    fun setParam(name: String, value: String) {
        params.putString(name, value)
    }

    fun getParam(name: String?): String? {
        return params.getString(name)
    }

    private fun addSpacePadding(string: String): String {
        return " $string "
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeBundle(params)
    }

    private constructor(parcel: Parcel) {
        params = parcel.readBundle(javaClass.classLoader) ?: Bundle()
    }

    companion object {
        private const val PARAM_QUERY = "QUERY"
        private const val PARAM_LANGUAGE = "LANGUAGE"

        @JvmField
        val CREATOR: Parcelable.Creator<Question> = object : Parcelable.Creator<Question> {
            override fun createFromParcel(source: Parcel): Question? {
                return Question(source)
            }

            override fun newArray(size: Int): Array<Question?> {
                return arrayOfNulls(size)
            }
        }
    }
}